src/merkle_tree.rs

```diff
@@ -1,27 +1,44 @@
-use std::hash::{DefaultHasher, Hash, Hasher};
+use std::rc::Rc;
+
+use sha3::Digest;
 
 pub enum SiblingHash {
-    Left(u64),
-    Right(u64),
+    Left(Vec<u8>),
+    Right(Vec<u8>),
+}
+
+#[derive(Clone)]
+struct InnerNode {
+    hash_value: Vec<u8>,
+    left_son: Rc<MerkleNode>,
+    right_son: Rc<MerkleNode>,
+}
+#[derive(Clone)]
+struct LeafNode {
+    hash_value: Vec<u8>,
 }
 
 #[derive(Clone)]
-struct MerkleNode {
-    hash_value: u64,
-    left_son: Option<Box<MerkleNode>>,
-    right_son: Option<Box<MerkleNode>>,
+enum MerkleNode {
+    Inner(InnerNode),
+    Leaf(LeafNode),
 }
-pub struct MerkleTree<H: Hash + Clone> {
+pub struct MerkleTree<H: AsRef<[u8]> + Clone> {
     merkle_root: MerkleNode,
-    leafs: Vec<H>,
+    leaves: Vec<H>,
 }
 
 impl MerkleNode {
-    pub fn new(
-        hash_value: u64,
-        left_son: Option<Box<MerkleNode>>,
-        right_son: Option<Box<MerkleNode>>,
-    ) -> Self {
+    pub fn get_hash_value(&self) -> Vec<u8> {
```

Comment:
Maybe you could return a `&[u8]` to avoid the `.clone()`.

Reply:
I've scaped enought, time to fight my biggest enemy, lifetimes (?